packages/nimble/R/genCpp_insertAssertions.R

```diff
@@ -31,19 +33,53 @@ exprClasses_insertAssertions <- function(code) {
                 toInsert <- lapply(code$args[[i]]$assertions, function(x) if(inherits(x, 'exprClass')) x else RparseTree2ExprClasses(x))
                 before <- unlist(lapply(toInsert, function(x) {if(x$name == 'after') FALSE else TRUE}))
                 if(nimbleOptions('experimentalNewSizeProcessing')) code$args[[i]]$assertions <- list()  ## Clear assertions field so it can be used by later compiler stages
-                newExpr <- newBracketExpr(args = c(lapply(toInsert[before], ## assertions will be inserted recursively IF they are in {}
-                                                          function(z) {exprClasses_insertAssertions(z); z}),
+                beforeContents <- lapply(toInsert[before], ## assertions will be inserted recursively IF they are in {}
+                                         function(z) {
+                                             exprClasses_insertAssertions(z)
+                                             z
+                                         })
+                afterContents <- lapply(toInsert[!before],
+                                        function(x) {
+                                            lapply(x$args[1],
+                                                   function(z) {
+                                                       exprClasses_insertAssertions(z)
+                                                   }
+                                                   )
+                                            x$args[[1]]
+                                        }
+                                        )
+                newExpr <- newBracketExpr(args = c(beforeContents,
                                                    code$args[i],
-                                                   lapply(toInsert[!before],
-                                                          function(x) {lapply(x$args[1],
-                                                                              function(z) {exprClasses_insertAssertions(z)}
-                                                                              ); x$args[[1]]}
-                                                          )))
+                                                   afterContents))
                 setArg(code, i, newExpr)
+                newExpr <- NULL
+            }
+            ## For 'if' or 'while', recurse into the if/while clause and (possibly) else clause.
+            ## 'while' causes a special challenge.  Any assertions must go before the condition
+            ## and also at the end of the clause for execution.  Otherwise the assertions will be
+            ## done only once on the first time through.
+            if(handleWhile) {
+                if(length(afterContents) > 0)
+                    stop('Invalid conditional in while()')
+                if(length(beforeContents) > 0) {
+                    ## The "while" is now embedded in newExpr, so it is in a different place in the AST.
```

Comment:
It's good to stand back and observe that this logic might be simpler in a [nanopass compiler](https://www.cs.indiana.edu/~dyb/pubs/nano-jfp.pdf), where there might be an entire isolated compiler stage devoted to rewriting
```r
while (complexExpression(x)) {
    doStuff()
}
```
to
```r
temp_123 <- complexExpression(x)
while (temp_123) {
    doStuff()
    temp_123 <- complexExpression(x)
}
```
Wouldn't that be easy to test!